install: Update if installed version != configured version

haslersn opened the pull request:
Previously, mattermost was never updated after its initial installation.
See #7.
    
Now we check the installed version and update mattermost if it doesn't
equal the configured version.
    
In https://docs.mattermost.com/administration/upgrade.html it's suggested
to keep certain folders when updating, so we keep those if they exist.

Closes #7.

haslersn committed:
install: Create a `version` file

The created file contains the `mattermost` version. This enables the
possibility to fix #7 in the future.

dasJ commented on tasks/install.yml:
```diff
@@ -36,3 +36,11 @@
     system: yes
     createhome: no 
+
+- name: Create version file
```
Please also check the contents of the file

haslersn replied:
Because of `force: yes`, the `version` file contents are overwritten every time. (I did also test this manually.)

SlothOfAnarchy replied:
I think he means that you should check the contents when installing or updating mattermost.

haslersn replied:
I added a corresponding commit and updated the PR description.

haslersn committed:
install: Update if installed version != configured version

Previously, mattermost was never updated after its initial installation.
See #7.

Now we check the installed version and update mattermost if it doesn't
equal the configured version.

In https://docs.mattermost.com/administration/upgrade.html it's suggested
to keep certain folders when updating, so we keep those if they exist.

SlothOfAnarchy commented on tasks/install.yml:
```diff
@@ -1,38 +1,114 @@
+    args:
+    when: mattermost_enterprise
+
+  when: mattermost_needs_update is defined
```
Please move these tasks to a new file e.g. `update_mattermost.yml` and include it with the condition

SlothOfAnarchy commented on tasks/install.yml:
```diff
@@ -1,38 +1,108 @@
+          item_path: /opt/mattermost/{{ item }}
+        with_items: '{{ mattermost_preserve_on_update }}'
+
+      - command: mv /opt/mattermost-displaced/{{ item }} /opt/mattermost/{{ item }}
```
I think this should be done after the update

haslersn replied:
I updated the PR.

SlothOfAnarchy commented on tasks/update.yml:
```diff
@@ -0,0 +1,91 @@
+
+- name: Remove a previously leaked mattermost-updated folder
+  file:
+    path: /opt/mattermost-updated
```
If this folder existed, it would have been removed using the earlier deploy task, so this task is unnecessary.

haslersn replied:
This is true in the context of `install.yml`, but I wanted `update.yml` to be self-contained. No matter of the previous state, running `update.yml` works and places the updated version inside `/opt/mattermost-updated/mattermost`.

SlothOfAnarchy commented on tasks/install.yml:
```diff
@@ -1,38 +1,30 @@
+
+- name: Set that mattermost needs to be updated
+  set_fact:
+    mattermost_needs_update:
```
You could also just set the variable's content to the condition. This way, you just need to check the variable and not whether it's defined

haslersn replied:
This is true in the context of `install.yml`, but I wanted `update.yml` to be self-contained. No matter of the previous state, running `update.yml` works and places the updated version inside `/opt/mattermost-updated/mattermost`.

SlothOfAnarchy commented on tasks/update.yml:
```diff
@@ -0,0 +1,91 @@
+
+- name: Create preserved folders at the destination
+  file:
+    path: /opt/mattermost-displaced/mattermost/{{ item }}
```
Is there a reason for the folder `mattermost`? Why not place the files directly in the displaced folder?

haslersn replied:
The folder inside the downloaded archive is called `mattermost` and it can't be renamed before/during extracting.

SlothOfAnarchy commented on tasks/update.yml:
```diff
@@ -0,0 +1,91 @@
+    path: /opt/mattermost-displaced/mattermost/{{ item }}
+    state: directory
+    recurse: yes
+  with_items: '{{ mattermost_preserve_on_update }}'
```
You could use the map and combine filters and only use one task for all folders

SlothOfAnarchy commented on tasks/update.yml:
```diff
@@ -0,0 +1,91 @@
+  with_items: '{{ mattermost_preserve_on_update }}'
+
+- name: Synchronize the preserved folders
+  command: rsync -aI /opt/mattermost/{{ item }}/ /opt/mattermost-displaced/mattermost/{{ item }}
```
Why not use the module for rsync, [synchronize](https://docs.ansible.com/ansible/2.5/modules/synchronize_module.html)?

haslersn replied:
It doesn't support any equivalent of `-I` (which is like force).

haslersn committed:
configure: Remove duplications

Three tasks were previously unnecessarily duplicated. Also, setting
users and groups prior to the task 'Change Mattermost directory
permissions' is redundant.
